test/e2e/httpproxy/namespaces_test.go

```diff
@@ -0,0 +1,216 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build e2e
+
+package httpproxy
+
+import (
+	"context"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	contourv1 "github.com/projectcontour/contour/apis/projectcontour/v1"
+	"github.com/projectcontour/contour/test/e2e"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func testWatchNamespaces(namespaces []string) e2e.NamespacedTestBody {
+	return func(nonWatchedNS string) {
+		Specify("HTTPProxies outside of watched namespaces are not configured", func() {
+			// Proxy in watched namespace should succeed
+			for _, ns := range namespaces {
+				deployEchoServer(f.T(), f.Client, ns, "echo")
+				p := newEchoProxy("proxy", ns)
+				f.CreateHTTPProxyAndWaitFor(p, e2e.HTTPProxyValid)
+
+				res, ok := f.HTTP.RequestUntil(&e2e.HTTPRequestOpts{
+					Host:      p.Spec.VirtualHost.Fqdn,
+					Condition: e2e.HasStatusCode(200),
+				})
+				require.NotNil(f.T(), res)
+				require.Truef(f.T(), ok, "expected 200 response code, got %d", res.StatusCode)
+			}
+
+			// Proxy in non-watched namespace should not be handled
+			deployEchoServer(f.T(), f.Client, nonWatchedNS, "echo")
+			p := newEchoProxy("proxy", nonWatchedNS)
+			err := f.CreateHTTPProxy(p)
+			require.Nil(f.T(), err, "could not create httpproxy")
+
+			// wait.Poll automatically waits the interval before first poll.
+			// If the resource is still NotReconsiled we can be fairly sure that the proxy is not handling it
+			interval := 10 * time.Second
+			ok := wait.Poll(interval, f.RetryTimeout, func() (bool, error) {
```

Comment:
We use `require.Never` in a few places for this sort of thing - poll periodically up to a time limit that a condition does not occur. In this case we could either check that it never becomes valid, or that it never loses the NotReconciled status. I have a slight preference for that approach for consistency.

Reply:
That's neat, thanks!